FirebaseAdmin/FirebaseAdmin/Messaging/InstanceIdClient.cs

```diff
@@ -0,0 +1,155 @@
+    /// </summary>
+    internal sealed class InstanceIdClient : IDisposable
+    {
+        private const string IidHost = "https://iid.googleapis.com";
+
+        private const string IidSubscriberPath = "iid/v1:batchAdd";
+
+        private const string IidUnsubscribePath = "iid/v1:batchRemove";
+
+        private readonly ConfigurableHttpClient httpClient;
+
+        private readonly HttpErrorHandler errorHandler;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InstanceIdClient"/> class.
+        /// </summary>
+        /// <param name="clientFactory">A default implentation of the HTTP client factory.</param>
+        /// <param name="credential">An instance of the <see cref="GoogleCredential"/> GoogleCredential class.</param>
+        public InstanceIdClient(HttpClientFactory clientFactory, GoogleCredential credential)
+        {
+            this.httpClient = clientFactory.ThrowIfNull(nameof(clientFactory))
+                .CreateAuthorizedHttpClient(credential);
+
+            this.errorHandler = new MessagingErrorHandler();
+        }
+
+        /// <summary>
+        /// Subscribes a list of registration tokens to a topic.
+        /// </summary>
+        /// <param name="topic">The topic name to subscribe to. /topics/ will be prepended to the topic name provided if absent.</param>
+        /// <param name="registrationTokens">A list of registration tokens to subscribe.</param>
+        /// <returns>A task that completes with a <see cref="TopicManagementResponse"/>, giving details about the topic subscription operations.</returns>
+        public async Task<TopicManagementResponse> SubscribeToTopicAsync(string topic, List<string> registrationTokens)
+        {
+            return await this.SendInstanceIdRequest(topic, registrationTokens, IidSubscriberPath).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Unsubscribes a list of registration tokens from a topic.
+        /// </summary>
+        /// <param name="topic">The topic name to unsubscribe from. /topics/ will be prepended to the topic name provided if absent.</param>
+        /// <param name="registrationTokens">A list of registration tokens to unsubscribe.</param>
+        /// <returns>A task that completes with a <see cref="TopicManagementResponse"/>, giving details about the topic unsubscription operations.</returns>
+        public async Task<TopicManagementResponse> UnsubscribeFromTopicAsync(string topic, List<string> registrationTokens)
+        {
+            return await this.SendInstanceIdRequest(topic, registrationTokens, IidUnsubscribePath).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Dispose the HttpClient.
+        /// </summary>
+        public void Dispose()
+        {
+            this.httpClient.Dispose();
+        }
+
+        private async Task<TopicManagementResponse> SendInstanceIdRequest(string topic, List<string> registrationTokens, string path)
```

Comment:
We need to validate `registrationTokens`:

1. Must not be null or empty
2. Length must not exceed 1000
3. Each entry in the list must not be null or empty